2004-11-30  Jim Meyering  <jim@meyering.net>

	* doio.c (Perl_io_close): Make the return value depend not only on the
	success of the close itself, but also on whether the output stream had
	a previous error.

diff -Naur --exclude=debian perl-5.8.4.orig/doio.c perl-5.8.4/doio.c
--- perl-5.8.4.orig/doio.c	2004-03-23 06:53:39.000000000 +1100
+++ perl-5.8.4/doio.c	2005-02-02 23:47:31.000000000 +1100
@@ -1014,11 +1014,14 @@
 	    retval = TRUE;
 	else {
 	    if (IoOFP(io) && IoOFP(io) != IoIFP(io)) {		/* a socket */
-		retval = (PerlIO_close(IoOFP(io)) != EOF);
+		bool prev_err = PerlIO_error(IoOFP(io));
+		retval = (PerlIO_close(IoOFP(io)) != EOF && !prev_err);
 		PerlIO_close(IoIFP(io));	/* clear stdio, fd already closed */
 	    }
-	    else
-		retval = (PerlIO_close(IoIFP(io)) != EOF);
+	    else {
+		bool prev_err = PerlIO_error(IoIFP(io));
+		retval = (PerlIO_close(IoIFP(io)) != EOF && !prev_err);
+	    }
 	}
 	IoOFP(io) = IoIFP(io) = Nullfp;
     }
